### PR TITLE
Send connection: close header

### PR DESCRIPTION
**Pull request by samphilipd**

### CODE REVIEW

File: app/lib/tokumei/close_connection.ex [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  end
+  defmacro __before_compile__(_env) do
+    quote location: :keep do
+      defoverridable [handle_request: 2]
+
+      def handle_request(request, env) do
+        response = super(request, env)
+
+        response_headers = case request.headers do
+          %{"connection" => "close"} -> [{"connection", "close"} | response.headers]
```
Comment (CrowdHailer) [inline app/lib/tokumei/close_connection.ex]: are request headers really a map? If so I have forgotten that change.

second should not add connection close if it is already in the response headers

Comment (samphilipd) [inline app/lib/tokumei/close_connection.ex]: What should it be if not a map?

Comment (CrowdHailer) [inline app/lib/tokumei/close_connection.ex]: is it not a list of tuple pairs? same as on the response?

Comment (samphilipd) [inline app/lib/tokumei/close_connection.ex]: According to your docs, Raxx.Response and Raxx.Request headers are both maps: 

https://hexdocs.pm/raxx/Raxx.Response.html

https://hexdocs.pm/raxx/Raxx.Request.html

Comment (CrowdHailer) [inline app/lib/tokumei/close_connection.ex]: Ok, but in this case you are matching on a map of headers and adding to a tuple list of response headers so one is incorrect. Also a little embarassing but I think my docs are lying to you. we chatted about this a lot on an issue previously https://github.com/CrowdHailer/raxx/issues/13. I would just put some inspects in somewhere and see what it really is

File: app/lib/tokumei/close_connection.ex [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        response = super(request, env)
+
+        response_headers = case request.headers do
+          %{"connection" => "close"} ->
+            if Enum.member?(response.headers , {"connection", "close"}) do
+              response.headers
+            else
+              [{"connection", "close"} | response.headers]
+            end
+          _ -> response.headers
```
Comment (CrowdHailer) [inline app/lib/tokumei/close_connection.ex]: new line between match and action please :whale:

File: app/test/tokumei/close_connection_test.exs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  use ExUnit.Case
+
+  alias Raxx.Response
+  alias Raxx.Request
+
+  use Tokumei.CloseConnection
+
+  def handle_request(%{method: :GET}, _), do: Response.ok("GET")
+
+  test "sends back connection: close if that was received" do
```
Comment (CrowdHailer) [inline app/test/tokumei/close_connection_test.exs]: Can we change these tests to run against running servers, connecting via a port? we can then test that the connection was closed as well as testing the headers are correct

Comment (samphilipd) [inline app/test/tokumei/close_connection_test.exs]: @CrowdHailer is there an example test I can follow?

Comment (CrowdHailer) [inline app/test/tokumei/close_connection_test.exs]: No don't think in this repo already. something like what is already set up in here is probably good https://github.com/CrowdHailer/Ace/blob/master/extensions/http(s)/test/ace/http/request_test.exs

File: app/lib/tokumei/close_connection.ex [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  defmacro __before_compile__(_env) do
+    quote location: :keep do
+      defoverridable [handle_request: 2]
+
+      def handle_request(request, env) do
+        response = super(request, env)
+
+        response_headers = case request.headers do
+          %{"connection" => "close"} ->
+            if Enum.member?(response.headers , {"connection", "close"}) do
```
Comment (CrowdHailer) [inline app/lib/tokumei/close_connection.ex]: If someone has explicitly set connection header to something else we should not override it

**Comment by CrowdHailer** [comment]
Sorry about this but thinking about it more I think that it belongs in Ace. A couple of reasons for this.

- HTTP/2 does not use the Connection header field to indicate connection-specific header fields
  https://tools.ietf.org/html/rfc7540#section-8.1.2.2

- talking to our ops guy about AWS load balancers the connection header is not normally forwarded through proxies.

For both those reasons applications often do not get exposed to the connection header at all. 
As the end goal for Ace is to normalise HTTP/1.1 and HTTP/2 I think ace_http should handle connection headers

tests should look like the gen_tcp based tests in this PR

- a request with `connection: close` should be replied to with a response with `connection: close`
- a `connection: close` header can be returned from the raxx application
- an application connection header will take priority over what is sent in.
- tests should check that the connection is closed even if the client does not close it
- the timeout should be configurable

Will reference the Ace PR once open

**Comment by samphilipd** [comment]
@CrowdHailer cool, I'd be interested to work on the Ace PR once the refactoring is done.

**Comment by CrowdHailer** [comment]
@samphilipd if you wanted a reasonably contained project you could tackle updating ace_http to the streaming version of Raxx. so then when the two sides are merged things should look the same. If that is of interest I could try and flesh out what needs to be done there so in can be tackled in a few not too crazy steps

**Comment by samphilipd** [comment]
@CrowdHailer sounds good, will that be on the ace project?

**Comment by CrowdHailer** [comment]
yes that will be work within the ace repo on the ace_http extension. should not require any changes to Raxx and Tokumei

**Comment by CrowdHailer** [comment]
@samphilipd Closing here we should recheck the problems you had with your odd client when next Ace.HTTP is released

**Pull request closed without merging**